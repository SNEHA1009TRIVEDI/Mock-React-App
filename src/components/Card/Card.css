.card-body-dark {
  width: 32%;
  height: 350px;
  background-color: #ababab;
  border-radius: 20px;
  border-color: black;
  box-shadow: 0 30px 40px rgba(0, 0, 0, 0.1);
  margin-bottom: 30px;
  text-overflow: ellipsis;
  box-shadow: rgba(0, 0, 0, 0.25) 0px 14px 28px,
    rgba(0, 0, 0, 0.22) 0px 10px 10px;
}
.card-body-light {
  width: 32%;
  height: 350px;
  background-color: #dfdfdf;
  border-radius: 20px;
  border-color: black;
  box-shadow: 0 30px 40px rgba(0, 0, 0, 0.1);
  margin-bottom: 30px;
  text-overflow: ellipsis;
  box-shadow: rgba(0, 0, 0, 0.25) 0px 14px 28px,
    rgba(0, 0, 0, 0.22) 0px 10px 10px;
}
/* .card-backgroud {
  background-color: #ababab;
} */

.heart {
  text-align: end;
  height: 40px;
  padding-right: 40px;
}
.count {
  text-align: start;
  padding-left: 20px;
  padding-bottom: 60px;
  position: relative;
  right: 2px;
  bottom: 38px;
  color: white;
}
.image {
  padding: 10px;
  padding-bottom: 10px;

  /* margin-bottom: 0px; */
}
.song-image {
  width: 75%;
  height: 75%;
  border-radius: 50%;
  border-color: black;
  border: solid;
  margin: 0px;
}
.lower-section {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
}

.song-left {
  padding-left: 30px;
  text-align: left;
  padding-bottom: 20px;
  text-overflow: ellipsis;
}
.song-name {
  text-overflow: ellipsis;
  overflow: hidden;
  display: -webkit-box;
  -webkit-line-clamp: 1;
  -webkit-box-orient: vertical;
  width: 100%;
}
