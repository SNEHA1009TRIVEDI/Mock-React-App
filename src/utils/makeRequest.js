import axios from 'axios'; 
const makeRequest = async (method,url) => {
  const data = await axios({
    method: method,
    url: url,
    headers: {
        "Content-Type": "application/json",
        "Accept": "application/json",
        "Authorization": "Bearer QWlzaHdhcnlhIE4"
        },

  });
  return data;
}
export default makeRequest;