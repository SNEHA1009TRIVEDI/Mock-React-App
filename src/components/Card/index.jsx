import React, { useEffect } from "react";
import "./Card.css";
// import image from '../../assets/genre-bollywood.png'
import heart from "../../assets/heart-red.svg";
import heart_grey from "../../assets/heart-gray.svg";
import makeRequest from "../../utils/makeRequest";
const Card = ({ data, cnt }) => {
  const { name, imageUrl, artist, genre } = data;
  const [isLiked, setIsLiked] = React.useState(false);
  const [count, setCount] = React.useState(0);
  useEffect(() => {
    makeRequest("GET", `http://localhost:8080/api/records/${data.id}/likes`)
      .then((res) => {
        setIsLiked(res.data.data.like);
        setCount(res.data.data.count);
      })
      .catch((err) => {
        console.log(err);
      });
  }, []);
  let image;
  

  const likeHandler = () => {
    if (isLiked) {
      makeRequest("PATCH", `http://localhost:8080/api/records/${data.id}/likes`,count,false)
        .then((res) => {
          setIsLiked(!isLiked);
          setCount(count - 1);
          // console.log("get request is made");
          // console.log(data.id);
        })
        .catch((err) => {
          console.log(err);
        });
    } else {
      makeRequest(
        "PATCH",
        `http://localhost:8080/api/records/${data.id}/likes`,
        count,
        true
      )
        .then((res) => {
          setIsLiked(!isLiked);
          setCount(count + 1);
          //  console.log("patch request is made");
          // // console.log(data.id);
        })
        .catch((err) => {
          console.log(err);
        });

      // setIsLiked(!isLiked);
      // setCount(count + 1);
    }
  };
  if (!isLiked) {
    image = heart_grey;
  }
  if (isLiked) {
    image = heart;
  }
  const color = cnt % 2 === 0 ? 'dark' : 'light';

  return (
    <div className={`card-body-${color}`}>
      <div className="image">
        <img className="song-image" src={imageUrl} alt="here is song image" />
      </div>
      <div className="lower-section">
        <div className="song-left">
          <h1 className="song-name">{name}</h1>
          <p>{artist.name}</p>
        </div>
        <div className="song-right">
          <img
            src={image}
            alt="this is heart image"
            className="heart"
            // onClick={likeHandler}
          />
          <p className="count" onClick={likeHandler}>{count}</p>
        </div>
      </div>
    </div>
  );
};

// export the component
export default Card;
