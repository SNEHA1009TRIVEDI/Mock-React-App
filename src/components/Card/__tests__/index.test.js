import { render, screen, waitFor, fireEvent } from "@testing-library/react";

import React from "react";
import makeRequest from "../../../utils/makeRequest";
import Card from "../../Card/index";

jest.mock("../../../utils/makeRequest");
const mockDataArr = {
  data: [
    {
      id: "0e2ddccf-f494-4c41-a499-90f8267f491a",
      name: "Beautiful Mistakes (feat. Megan Thee Stallion)",
      genre: {
        id: "128aa7f8-c943-48ce-b352-7edd26fa4c6e",
        name: "Pop",
      },
      artist: {
        id: "496b0a85-2bfa-45bc-8d0f-57fe0ce55708",
        name: "Maroon 5",
      },
      imageUrl:
        "https://i.scdn.co/image/ab67616d0000b27386a8ab515de4b7aef28cd631",
      publishedAt: "2021-06-11T00:00:00",
    },
  ],
};

const mockData = {
  data: {
    data: {
      like: false,
      count: 15,
    },
  },
};

describe("Card", () => {
  it("should change the color of heart from grey to red", async () => {
    makeRequest.mockResolvedValue(mockData);
    const { getByTestId } = render(<Card data={mockDataArr.data[0]} cnt={1} />);
    const element = getByTestId("heart-image");
    // console.log(element.src);
    expect(element.src).toContain("heart-gray.svg");
    fireEvent.click(element);
    await waitFor(() => {
      expect(element.src).toContain("heart-red.svg");
    });
  });
  it("should change the color of heart from red to grey", async () => {
    makeRequest.mockResolvedValue(mockData);
    const { getByTestId } = render(<Card data={mockDataArr.data[0]} cnt={2} />);
    const element = getByTestId("heart-image");
    fireEvent.click(element);
    // console.log(element.src);
    fireEvent.click(element);
    await waitFor(() => {
      expect(element.src).toContain("heart-red.svg");
    });
  });
  it("shoukd display the count value", async () => {
    makeRequest.mockResolvedValue(mockData);
    const { getByTestId } = render(<Card data={mockDataArr.data[0]} cnt={2} />);
    const count = getByTestId("count");
    await waitFor(() => {
      expect(count.textContent).toBe("15");
    });
  });
  it("should render the card correctly", async () => {
    makeRequest.mockResolvedValue(mockData);
    const { container } = render(<Card data={mockDataArr.data[0]} cnt={2} />);
    await waitFor(() => {
      expect(container).toMatchSnapshot();
    });
  });
});
