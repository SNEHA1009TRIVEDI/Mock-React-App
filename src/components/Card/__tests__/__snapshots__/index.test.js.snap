// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card should render the card correctly 1`] = `
<div>
  <div
    class="card-body-dark"
  >
    <div
      class="image"
    >
      <img
        alt="here is song image"
        class="song-image"
        src="https://i.scdn.co/image/ab67616d0000b27386a8ab515de4b7aef28cd631"
      />
    </div>
    <div
      class="lower-section"
    >
      <div
        class="song-left"
      >
        <h1
          class="song-name"
        >
          Beautiful Mistakes (feat. Megan Thee Stallion)
        </h1>
        <p>
          Maroon 5
        </p>
      </div>
      <div
        class="song-right"
      >
        <img
          alt="this is heart image"
          class="heart"
          data-testid="heart-image"
          src="heart-gray.svg"
        />
        <p
          class="count"
          data-testid="count"
        >
          0
        </p>
      </div>
    </div>
  </div>
</div>
`;
